colexec: add buffering stage to hash aggregator

yuzefovich opened the pull request:
This commit adds a buffering stage to the hash aggregator in which we
accumulate tuples from the input until we get `coldata.MaxBatchSize` of
them and only then we proceed to the online aggregation. This shows nice
performance improvements in most of the cases due to the fact that with
more tuples in the online aggregation we're more likely to have more
tuples that belong to the same group, so we end up processing all of
them at once rather that in separate batches.

One thing worth calling out is that `appendOnlyBufferedBatch` has been
extended to support a selection vector of an arbitrary length. This is
necessary because online aggregation updates the selection vector on the
buffered tuples several times.

Release justification: low risk, high benefit change.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53380)
<!-- Reviewable:end -->

yuzefovich commented:
The benchmarks are quite [good](https://gist.github.com/yuzefovich/931aed43b8767be6a373f545ff082993).

asubiotto reviewed (changes requested):
Didn't we already have a buffering stage for online aggregation? I remember Archer ran some benchmarks to figure out a good value for the number of tuples to buffer.

I'm also not convinced that this change should go in during the stability period, but let's discuss during standup.

> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-:-MG2yVUL7j06Gc18x4oQ:bcpz4wq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/hash_aggregator.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-MG2zlAg8di9fdXhd3Ru:-MG2zlAhDh3lmuMj48mB:b996585) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3d28c9c7bed7c0ad7f76be96b8f555c066cef6/pkg/sql/colexec/hash_aggregator.go#L130)):*
> ```Go
> // coldata.MaxBatchSize, but then we would lose some test coverage due to
> // disabling of the randomization of the batch size, so we, instead, use
> // 4 x coldata.BatchSize() (which ends up being coldata.MaxBatchSize in
> ```

I'm not sure I understand why using a larger batch size increases test coverage because it's still fixed, right? Do you want to add a way to modify this in tests?
___
*[pkg/sql/colexec/hash_aggregator.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-MG2yA5-BxpEvkoi151m:-MG2yA5-BxpEvkoi151n:bgj7sj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3d28c9c7bed7c0ad7f76be96b8f555c066cef6/pkg/sql/colexec/hash_aggregator.go#L132)):*
> ```Go
> // 4 x coldata.BatchSize() (which ends up being coldata.MaxBatchSize in
> // non-test environment).
> var hashAggregatorMaxBuffered = 4 * coldata.BatchSize()
> ```

Why not make this `coldata.MaxBatchSize` or some multiple of that?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Yes, we did have a buffering stage, but it was later removed for optimization purposes. However, since we have refactored the algorithm of hash aggregation it now became better to add the buffering stage back. At that time, 8192 was the number of buffered tuples, but those results aren't really applicable here since the algorithm has changed fundamentally.

We decided to hold off on merging this until the branch is cut.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-:-MG4NwqgCTDPUjsdjto7:b-n6gyx8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/hash_aggregator.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-MG2zlAg8di9fdXhd3Ru:-MG4OUbi2G5V1i4kdJTR:bq2qbdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3d28c9c7bed7c0ad7f76be96b8f555c066cef6/pkg/sql/colexec/hash_aggregator.go#L130)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm not sure I understand why using a larger batch size increases test coverage because it's still fixed, right? Do you want to add a way to modify this in tests?
</blockquote></details>

No, I think you're missing the point here: according to the benchmarks, the larger this number is, the better the speed is. In several places throughout the engine we make an assumption that `coldata.MaxBatchSize` is the maximum allocated length of things (for example, in the hash table), so we cannot go larger without modifications elsewhere. So we want `hashAggregatorMaxBuffered` to be `coldata.MaxBatchSize`; however, as I replied below, if we do so, we're losing some of the test coverage.

I don't want to introduce any others knobs because I don't like having custom knobs when we're randomizing `coldata.BatchSize()` - I think we should try to take advantage of that randomization here.
___
*[pkg/sql/colexec/hash_aggregator.go, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-MG2yA5-BxpEvkoi151m:-MG4O0ryEGzZnVEHInPE:b-ofuqux) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3d28c9c7bed7c0ad7f76be96b8f555c066cef6/pkg/sql/colexec/hash_aggregator.go#L132)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why not make this `coldata.MaxBatchSize` or some multiple of that?
</blockquote></details>

Because then we're losing the randomization of `coldata.BatchSize()` - what ends up happening is that `hashAggregatorMaxBuffered` is initialized to a constant value, but we might call `SetBatchSizeForTests` later, still `hashAggregatorMaxBuffered` doesn't change, so we're losing the test coverage in our unit tests.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53380)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/colexec/hash_aggregator.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-MG2zlAg8di9fdXhd3Ru:-MG7kTog-V-lzbFCZAhS:b-w73k4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3d28c9c7bed7c0ad7f76be96b8f555c066cef6/pkg/sql/colexec/hash_aggregator.go#L130)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

No, I think you're missing the point here: according to the benchmarks, the larger this number is, the better the speed is. In several places throughout the engine we make an assumption that `coldata.MaxBatchSize` is the maximum allocated length of things (for example, in the hash table), so we cannot go larger without modifications elsewhere. So we want `hashAggregatorMaxBuffered` to be `coldata.MaxBatchSize`; however, as I replied below, if we do so, we're losing some of the test coverage.

I don't want to introduce any others knobs because I don't like having custom knobs when we're randomizing `coldata.BatchSize()` - I think we should try to take advantage of that randomization here.
</blockquote></details>

But isn't `coldata.BatchSize` not random at this point? (i.e. the `init` function that randomizes this won't have run)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53380)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/hash_aggregator.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-MG2zlAg8di9fdXhd3Ru:-MG9EGYd4hDW_mJ9jwJo:b-jof8pm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc3d28c9c7bed7c0ad7f76be96b8f555c066cef6/pkg/sql/colexec/hash_aggregator.go#L130)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

But isn't `coldata.BatchSize` not random at this point? (i.e. the `init` function that randomizes this won't have run)
</blockquote></details>

Yes, you're totally right, my bad, fixed. In my original version I was setting this variable in `NewHashAggregator`, and at that time `coldata.BatchSize()` is randomized, but then I incorrectly refactored it without giving it much thought.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: add buffering stage to hash aggregator

This commit adds a buffering stage to the hash aggregator in which we
accumulate tuples from the input until we get `coldata.MaxBatchSize` of
them and only then we proceed to the online aggregation. This shows nice
performance improvements in most of the cases due to the fact that with
more tuples in the online aggregation we're more likely to have more
tuples that belong to the same group, so we end up processing all of
them at once rather that in separate batches.

One thing worth calling out is that `appendOnlyBufferedBatch` has been
extended to support a selection vector of an arbitrary length. This is
necessary because online aggregation updates the selection vector on the
buffered tuples several times.

Release justification: low risk, high benefit change.

Release note: None

yuzefovich commented:
Updated the PR to de-risk it as we discussed.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53380#-:-MGEZaM18qqearHRzw1V:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
